src/Stratis.Bitcoin.Features.Consensus/CoinViews/DBreezeCoinView.cs

```diff
@@ -293,44 +296,72 @@ public RewindData GetRewindData(int height)
+
+                if (targetHeight <= 0)
                 {
                     transaction.RemoveAllKeys("Coins", true);
-                    this.SetBlockHash(transaction, this.network.GenesisHash);
+                    transaction.RemoveAllKeys("Rewind", true);
 
                     res = this.network.GenesisHash;
                 }
                 else
                 {
-                    transaction.ValuesLazyLoadingIsOn = false;
-
-                    Row<int, byte[]> firstRow = transaction.SelectBackward<int, byte[]>("Rewind").FirstOrDefault();
-                    transaction.RemoveKey("Rewind", firstRow.Key);
-                    var rewindData = this.dBreezeSerializer.Deserialize<RewindData>(firstRow.Value);
-                    this.SetBlockHash(transaction, rewindData.PreviousBlockHash);
+                    var keysToRemove = new List<int>();
+                    var changes = new Dictionary<uint256, Coins>();
 
-                    foreach (uint256 txId in rewindData.TransactionsToRemove)
+                    while (currentHeight > targetHeight)
                     {
-                        this.logger.LogTrace("Outputs of transaction ID '{0}' will be removed.", txId);
-                        transaction.RemoveKey("Coins", txId.ToBytes(false));
-                    }
+                        transaction.ValuesLazyLoadingIsOn = false;
 
-                    foreach (UnspentOutputs coin in rewindData.OutputsToRestore)
-                    {
-                        this.logger.LogTrace("Outputs of transaction ID '{0}' will be restored.", coin.TransactionId);
-                        transaction.Insert("Coins", coin.TransactionId.ToBytes(false), this.dBreezeSerializer.Serialize(coin.ToCoins()));
+                        Row<int, byte[]> firstRow = transaction.Select<int, byte[]>("Rewind", currentHeight);
+                        var rewindData = this.dBreezeSerializer.Deserialize<RewindData>(firstRow.Value);
+                        keysToRemove.Add(firstRow.Key);
+
+                        foreach (uint256 txId in rewindData.TransactionsToRemove)
+                        {
+                            this.logger.LogTrace("Outputs of transaction ID '{0}' will be removed.", txId);
+                            changes[txId] = null;
+                        }
+
+                        foreach (UnspentOutputs coin in rewindData.OutputsToRestore)
+                        {
+                            this.logger.LogTrace("Outputs of transaction ID '{0}' will be restored.", coin.TransactionId);
+                            changes[coin.TransactionId] = coin.ToCoins();
+                        }
+
+                        res = rewindData.PreviousBlockHash;
+
+                        currentHeight--;
                     }
 
-                    res = rewindData.PreviousBlockHash;
+                    var byteListComparer = new ByteListComparer();
+
+                    foreach (int key in keysToRemove.OrderBy(k => k))
```

Comment:
.Reverse will produce the same order

Reply:
The idea here is to remove keys in ascending order for DBreeze optimization purposes.